src/transformState.js

```diff
@@ -5,7 +5,33 @@
  * @param {Object[]} actions
  */
 function transformState(state, actions) {
-  // write code here
+  for (let i = 0; i < actions.length; i++) {
+    const action = actions[i];
+
+    if (action.type === 'addProperties') {
+      const extraData = action.extraData;
+
+      for (const key in extraData) {
+        state[key] = extraData[key];
+      }
+    } else if (action.type === 'removeProperties') {
+      const keysToRemove = action.keysToRemove;
+
+      for (let j = 0; j < keysToRemove.length; j++) {
+        const key = keysToRemove[j];
+
+        if (state.hasOwnProperty(key)) {
+          delete state[key];
+        }
+      }
+    } else if (action.type === 'clear') {
+      for (const key in state) {
+        delete state[key];
+      }
+    }
+  }
+
+  return state;
```

Comment:
The function should not return the state as it is modified by reference. Remove the return statement.